pkg/cache/v3/linear.go

```diff
@@ -634,71 +524,92 @@ func (cache *LinearCache) CreateDeltaWatch(request *DeltaRequest, sub Subscripti
 
 	watch := DeltaResponseWatch{Request: request, Response: value, subscription: sub}
 
+	// On first request on a wildcard subscription, envoy does expect a response to come in to
+	// conclude initialization.
+	replyEvenIfEmpty := sub.IsWildcard() && request.GetResponseNonce() == ""
+
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
-	response, err := cache.computeDeltaResponse(watch)
+	response, err := cache.computeResponse(watch, replyEvenIfEmpty)
 	if err != nil {
 		return nil, fmt.Errorf("failed to compute the watch respnse: %w", err)
 	}
 
 	if response != nil {
 		cache.log.Debugf("[linear cache] replying to the delta watch (subscription values %v, known %v)", sub.SubscribedResources(), sub.ReturnedResources())
-		watch.Response <- response
+		watch.sendResponse(response)
 		return nil, nil
 	}
 
+	return cache.trackWatch(watch), nil
+}
+
+func (cache *LinearCache) nextWatchID() uint64 {
+	cache.currentWatchID++
+	if cache.currentWatchID == 0 {
+		panic("watch id count overflow")
+	}
+	return cache.currentWatchID
+}
+
+// Must be called under lock
+func (cache *LinearCache) trackWatch(watch watch) func() {
+	cache.watchCount++
+
 	watchID := cache.nextWatchID()
-	// Create open watches since versions are up to date.
+	sub := watch.getSubscription()
+
 	if sub.IsWildcard() {
-		cache.log.Infof("[linear cache] open delta watch %d for all %s resources, system version %q", watchID, cache.typeURL, cache.getVersion())
-		cache.wildcardWatches.delta[watchID] = watch
+		cache.log.Infof("[linear cache] open watch %d for %s all resources", watchID, cache.typeURL)
+		cache.log.Debugf("[linear cache] subscription details for watch %d: known versions %v, system version %q", watchID, sub.ReturnedResources(), cache.getVersion())
```

Comment:
would it be worth adding whether the watch was delta or not here? I know the goal is to completely abstract this away from the caches but I think it might help debug potential bugs in the short term considering the amount of changes that have been recently introduced.